### PR TITLE
docker-compose with nginx reverse proxy and letsencrypt certbot

### PR DESCRIPTION
**Pull request by daN4cat**
Here the docker-compose enhancement to add a nginx reverse proxy with SSL termination based on a Letsencrypt certificate.
The certbot docker image manages the SSL certificate refresh every 3 months.

To run locally you need to use the init-selfcert.sh script.
If you want to try the Letsencrypt SSL, you need to use the init-letsencrypt.sh script.

Please note that myPhpAdmin is enabled on port 8000. If not required the nginx forward rule needs to be commented out.

I didn't touch the docker-compose.dev and .test as I'm not sure how those are used.

@jekkos the issue I had was fixed setting FORCE_HTTPS=true in the environment: docker-compose.yml file.

### CODE REVIEW

File: init-letsencrypt.sh [hunk 1]
```diff
@@ -0,0 +1,75 @@
+#!/bin/bash
+
```
Comment (jekkos) [inline init-letsencrypt.sh]: maybe move these scripts to a subdirectory as the project root already contains quite some files?

**Comment by jekkos** [comment]
Thanks for this useful addition. One remark is that maybe we should move some of the init scripts to a separate folder (or reorganize the files in the project root a bit).

Does it renew the certificate everytime the container starts or is the cert persisted after it's initialized?

**Comment by jekkos** [comment]
Also related to #1139

**Comment by jekkos** [comment]
We might also want to add some documentation on this setup in the wiki after it's merged.

File: data/nginx/nginx.conf [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+		}
+
+		location / {
+			return 301 https://$host$request_uri;
+		}
+	}
+
+	server {
+		listen 443 ssl;
+		server_name ospos.ospos;
```
[the remark this answers is not in the document]
Comment (daN4cat) [reply data/nginx/nginx.conf]: Yes, something to parametrise

File: docker-compose.yml [hunk 3]
```diff
@@ -51,17 +53,50 @@ services:
...
+          - ospos
+          - phpmyadmin
+          - certbot
+        volumes:
+          - ./data/nginx/nginx.conf:/etc/nginx/nginx.conf:ro
+          - ./data/nginx/error_log.log:/etc/nginx/error_log.log
+          - ./data/certbot/conf:/etc/letsencrypt
+          - ./data/certbot/www:/var/www/certbot
+        ports:
+          - "80:80"
```
[the remark this answers is not in the document]
Comment (daN4cat) [reply docker-compose.yml]: Required by Letsencrypt to read the ACME file. Only that as the rest is redirected to https

File: init-letsencrypt.sh (hunk 1 above)
Comment (daN4cat) [reply init-letsencrypt.sh]: Yes, I believe it's better to move everything to a folder and reorg the docker setup.

File: init-letsencrypt.sh [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  curl -s https://raw.githubusercontent.com/certbot/certbot/master/certbot-nginx/certbot_nginx/options-ssl-nginx.conf > "$data_path/conf/options-ssl-nginx.conf"
+  curl -s https://raw.githubusercontent.com/certbot/certbot/master/certbot/ssl-dhparams.pem > "$data_path/conf/ssl-dhparams.pem"
+  echo
+fi
+
+echo "### Creating dummy certificate for $domains ..."
+path="/etc/letsencrypt/live/$domains"
+mkdir -p "$data_path/conf/live/$domains"
+docker-compose run --rm --entrypoint "\
+  openssl req -x509 -nodes -newkey rsa:1024 -days 1\
```
[the remark this answers is not in the document]
Comment (daN4cat) [reply init-letsencrypt.sh]: This is just the temp cert, right?

**Comment by daN4cat** [comment]
Yes, it needs some polishing. This is just functioning and a draft for discussion.

The LEMP stack means we would need to change ospos from Apache to nginx. Something I can do as second step.
This is more a reverse proxy to terminate SSL with free cert and auto renewal.

I took some of the docker code from GitHub, there might be other ways to solve the same problem.

**Comment by daN4cat** [comment]
The cert is initially created in a manual way and signed, it stays in a local folder. But I need to countercheck it.

And yes about the wiki documentation.

**Comment by daN4cat** [comment]
@jekkos see if the new setup looks better.

All Docker business is now in a docker folder.

From docker folder:
- for local setup run the script install-local.sh
- for server setup run the script install-server.sh **
- to tidy up run the script uninstall.sh

** to configure all the parameters change the file .env, especially if you run this on a server in the public Internet!

I didn't touch the .dev and .test files because I don't know how those are used. It's probably better if you update them.

**Comment by jekkos** [comment]
@daN4cat I meant that moving only the init scripts to a subfolder would be enough. It might be rather hard to move the Dockerfiles themselves as well (as then you break all references from inside the mounted paths in the container itself). Most of the projects do have those in their root folder anyway.

**Comment by jekkos** [comment]
.dev and .test files are not much different from an actual container setup. They assume that you mount your working directory into the container which is handy for development. It also adds an extra xdebug package that shouldn't be installed in a production setup. I might want to move this into the regular file by adding a flag read from an env variable instead for example.

**Comment by daN4cat** [comment]
Indeed, I need to move Docker files back to the root.
The mounting of folder is a complication.

**Comment by daN4cat** [comment]
I don't need to touch that much dev and test, you don't need the nginx reverse proxy for that.
I might look into touching a bit the networking, ports and the images but not too much else.

**Comment by objecttothis** [comment]
> The LEMP stack means we would need to change ospos from Apache to nginx. Something I can do as second step.
I do think we need to retain general Apache compatibility for OSPOS.

File: data/nginx/nginx.conf (hunk 2 above)
[the remark this answers is not in the document]
Comment (daN4cat) [reply data/nginx/nginx.conf]: This is done, the .env variable is used to fill in the server name

File: init-letsencrypt.sh (hunk 1 above)
Comment (daN4cat) [reply init-letsencrypt.sh]: script moved to docker subfolder.

In order to do install this, from local do:

> sudo docker/install-local.sh

for a server case (remember to change the docker/.env values!!)

> sudo docker/install-server.sh

**Comment by daN4cat** [comment]
@jekkos do you want to give it a try on your server before I merge this?

File: Dockerfile [hunk 5]
```diff
@@ -1,5 +1,6 @@
-FROM php:7.2.18-apache
+FROM php:7.2-apache
```
Comment (jekkos) [inline Dockerfile]: we might want to stick with 7.2.18 here (latest)

Comment (daN4cat) [reply Dockerfile]: I was under the impression that the docker image with this generic version format will pick up the latest.
Also I may bumped it to be 7.3 so we test the latest.

Comment (daN4cat) [reply Dockerfile]: Upgraded to 7.3

**Comment by daN4cat** [comment]
To use this new Docker setup.

From ospos root folder:
- > chmod ugo+x docker/*.sh
- for local setup run the script: > docker/install-local.sh (say y everytime you are asked to renew the cert)
- for server setup run the script: > docker/install-server.sh **
- to tidy up run the script: > docker/uninstall.sh

** to configure all the parameters change the file .env, especially if you run this on a server in the public Internet!
Also make sure you close phpMyAdmin if you don't want to expose that to the public internet.

**Comment by daN4cat** [comment]
@jekkos are you fine for me to merge this?

**Pull request merged**